src/_avif.c

```diff
@@ -0,0 +1,819 @@
+    }
+
+    avifRGBImageFreePixels(&rgb);
+
+    Py_RETURN_NONE;
+}
+
+PyObject *
+_encoder_finish(AvifEncoderObject *self) {
+    avifEncoder *encoder = self->encoder;
+
+    avifRWData raw = AVIF_DATA_EMPTY;
+    avifResult result;
+    PyObject *ret = NULL;
+
+    Py_BEGIN_ALLOW_THREADS
+    result = avifEncoderFinish(encoder, &raw);
+    Py_END_ALLOW_THREADS
+
+    if (result != AVIF_RESULT_OK) {
+        PyErr_Format(
+            exc_type_for_avif_result(result),
+            "Failed to finish encoding: %s",
+            avifResultToString(result));
+        avifRWDataFree(&raw);
+        return NULL;
+    }
+
+    ret = PyBytes_FromStringAndSize((char *)raw.data, raw.size);
+
+    avifRWDataFree(&raw);
+
+    return ret;
+}
+
+// Decoder functions
+PyObject *
+AvifDecoderNew(PyObject *self_, PyObject *args) {
+    const uint8_t *avif_bytes;
+    Py_ssize_t size;
+    AvifDecoderObject *self = NULL;
+
+    char *upsampling_str;
+    char *codec_str;
+    avifCodecChoice codec;
+    avifChromaUpsampling upsampling;
+
+    avifResult result;
+
+    if (!PyArg_ParseTuple(
+            args, "z#ss", &avif_bytes, &size, &codec_str, &upsampling_str)) {
+        return NULL;
+    }
+
+    if (!strcmp(upsampling_str, "auto")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_AUTOMATIC;
+    } else if (!strcmp(upsampling_str, "fastest")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_FASTEST;
+    } else if (!strcmp(upsampling_str, "best")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_BEST_QUALITY;
+    } else if (!strcmp(upsampling_str, "nearest")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_NEAREST;
+    } else if (!strcmp(upsampling_str, "bilinear")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_BILINEAR;
+    } else {
+        PyErr_Format(PyExc_ValueError, "Invalid upsampling option: %s", upsampling_str);
+        return NULL;
+    }
+
+    if (strcmp(codec_str, "auto") == 0) {
+        codec = AVIF_CODEC_CHOICE_AUTO;
+    } else {
+        codec = avifCodecChoiceFromName(codec_str);
+        if (codec == AVIF_CODEC_CHOICE_AUTO) {
+            PyErr_Format(PyExc_ValueError, "Invalid codec: %s", codec_str);
+            return NULL;
+        } else {
+            const char *codec_name = avifCodecName(codec, AVIF_CODEC_FLAG_CAN_DECODE);
+            if (codec_name == NULL) {
+                PyErr_Format(
+                    PyExc_ValueError, "AV1 Codec cannot decode: %s", codec_str);
+                return NULL;
+            }
+        }
+    }
+
+    self = PyObject_New(AvifDecoderObject, &AvifDecoder_Type);
+    if (!self) {
+        PyErr_SetString(PyExc_RuntimeError, "could not create decoder object");
+        return NULL;
+    }
+    self->decoder = NULL;
+    self->size = size;
+    self->data = PyMem_New(uint8_t, size);
+    if (self->data == NULL) {
+        PyErr_SetString(PyExc_MemoryError, "PyMem_New() failed");
+        Py_DECREF(self);
+        return NULL;
+    }
+
+    memcpy(self->data, avif_bytes, size);
```

Comment:
Document here as well.

Reply:
I wasn't entirely sure what you wanted documented for this line. I added this, let me know if it's what you had in mind:

https://github.com/python-pillow/Pillow/blob/b84a8e0d47d9e8807d207aba9c9eeb455646ef62/src/_avif.c#L484-L485

Reply:
I was able to avoid a memcpy here by having `PyArg_ParseTuple` pass in a `PyBytesObject` and incrementing the reference in the new / decrementing in the dealloc. That also avoids an unnecessary malloc during decoding.

Reply:
Realized it would probably be better to have you resolve these conversations, to confirm that the feedback has indeed been addressed.